### PR TITLE
Added limit/offset clause

### PR DESCRIPTION
**Pull request by beikov**

### CODE REVIEW

File: src/main/antlr/org/hibernate/sqm/parser/hql/internal/antlr/HqlParser.g4 [hunk 1]
```diff
@@ -66,6 +66,7 @@ targetFieldsSpec
 // ORDER BY clause
 
 orderByClause
+// todo : null precedence
```
Comment (sebersole) [inline src/main/antlr/org/hibernate/sqm/parser/hql/internal/antlr/HqlParser.g4]: null-precedence really belongs on `sortSpecification` rule, not `orderByClause`.

File: src/main/antlr/org/hibernate/sqm/parser/hql/internal/antlr/HqlParser.g4 [hunk 2]
```diff
@@ -87,10 +88,26 @@ orderingSpecification
...
+	;
+
+offsetClause
+	: OFFSET parameterOrNumberLiteral
+	;
+
+parameterOrNumberLiteral
+	: parameter
+	| INTEGER_LITERAL
+	;
```
Comment (sebersole) [inline src/main/antlr/org/hibernate/sqm/parser/hql/internal/antlr/HqlParser.g4]: I'm not sure we want to support parameters here.  Not all databases support parameterized limit/offset, even if they support limit/offset at all.

Comment (beikov) [inline src/main/antlr/org/hibernate/sqm/parser/hql/internal/antlr/HqlParser.g4]: If they don't support it, I could inline the value. Would that be ok?

Comment (beikov) [inline src/main/antlr/org/hibernate/sqm/parser/hql/internal/antlr/HqlParser.g4]: That obviously requires deeper integration but I think parameters might be nice here. I am imagining to add the limit/offset parameter values to the query cache key if the dbms doesn't support parameters. wdys?

Comment (sebersole) [inline src/main/antlr/org/hibernate/sqm/parser/hql/internal/antlr/HqlParser.g4]: In regards to the query-plan cache... If we support parameters here, IMO in cases where the database does not support that we should simply not cache the query plan.  I already have a bunch of checks for cases where we should not be caching the query plans such as multi-values parameters, etc.

However, I am not convinced we should support parameters here.  @dreab8, @Naros What are y'alls thoughts on this?

At the minimum, if the user specifies LIMIT/OFFSET using parameters and the database does not support that we should give a big fat warning.

Comment (beikov) [inline src/main/antlr/org/hibernate/sqm/parser/hql/internal/antlr/HqlParser.g4]: I'm more +1 for a warning than not being able to make use of parameters. As far as I know the most widespread DBs, all of them allow parameters. I'm fine with not caching in that case, but I would suggest that being configurable via a property.

Comment (dreab8) [inline src/main/antlr/org/hibernate/sqm/parser/hql/internal/antlr/HqlParser.g4]: I do not have a strong opinion, probably a warn and not caching is fine

**Comment by sebersole** [comment]
I applied this upstream

**Pull request closed without merging**